libbpf-tools/fsslower.c

```diff
@@ -0,0 +1,463 @@
+{
+	const struct event *e = data;
+	struct tm *tm;
+	char ts[32];
+	time_t t;
+
+	if (csv) {
+		printf("%lld,%s,%d,%c,", e->end_ns, e->task, e->pid, file_op[e->op]);
+		e->size == LLONG_MAX ? printf("LL_MAX,") : printf("%ld,", e->size);
+		printf("%lld,%lld,%s\n", e->offset, e->delta_us, e->file);
+		return;
+	}
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+
+	printf("%-8s %-16s %-7d %c ", ts, e->task, e->pid, file_op[e->op]);
+	e->size == LLONG_MAX ? printf("%-7s ", "LL_MAX") : printf("%-7ld ", e->size);
+	printf("%-8lld %7.2f %s\n", e->offset / 1024, (double)e->delta_us / 1000, e->file);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct fsslower_bpf *skel;
+	__u64 time_end = 0;
+	int err;
+	bool support_fentry;
+
+	alias_parse(argv[0]);
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+	if (fs_type == NONE) {
+		warn("filesystem must be specified using -t option.\n");
+		return 1;
+	}
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		warn("failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	skel = fsslower_bpf__open();
+	if (!skel) {
+		warn("failed to open BPF object\n");
+		return 1;
+	}
+
+	skel->rodata->target_pid = target_pid;
+	skel->rodata->min_lat_ms = min_lat_ms;
+
+	/*
+	 * before load
+	 * if fentry is supported, we set attach target and disable kprobes
+	 * otherwise, we disable fentry and attach kprobes after loading
+	 */
+	support_fentry = check_fentry();
+	if (support_fentry) {
+		err = fentry_set_attach_target(skel);
+		if (err) {
+			warn("failed to set attach target: %d\n", err);
+			goto cleanup;
+		}
+		disable_kprobes(skel);
+	} else {
+		disable_fentry(skel);
+	}
+
+	err = fsslower_bpf__load(skel);
+	if (err) {
+		warn("failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	/*
+	 * after load
+	 * if fentry is supported, let libbpf do auto load
+	 * otherwise, we attach to kprobes manually
+	 */
+	err = support_fentry ? fsslower_bpf__attach(skel) : attach_kprobes(skel);
+	if (err) {
+		warn("failed to attach BPF programs: %d\n", err);
+		goto cleanup;
+	}
+
+	signal(SIGINT, sig_handler);
+
+	pb_opts.sample_cb = handle_event;
+	pb_opts.lost_cb = handle_lost_events;
+	pb = perf_buffer__new(bpf_map__fd(skel->maps.events), PERF_BUFFER_PAGES,
+			      &pb_opts);
+	err = libbpf_get_error(pb);
+	if (err) {
+		pb = NULL;
+		warn("failed to open perf buffer: %d\n", err);
+		goto cleanup;
+	}
+
+	print_headers();
+
+	if (duration)
+		time_end = get_ktime_ns() + duration * NSEC_PER_SEC;
+
+	while (1) {
+		usleep(PERF_BUFFER_TIME_MS * 1000);
```

Comment:
Why is this usleep() needed here? I see that xfsslower is doing the same, but I can't recall why it was added in the first place...